store/localstore/local_region.go

```diff
@@ -552,36 +561,49 @@ func getIndexRowFromRange(idxInfo *tipb.IndexInfo, txn kv.Transaction, ran kv.Ke
 				break
 			}
 		}
-
-		datums, err := tablecodec.DecodeIndexKey(it.Key())
+		values, b, err := tablecodec.CutIndexKey(it.Key(), ids)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		var handle types.Datum
-		if len(datums) > len(idxInfo.Columns) {
-			handle = datums[len(idxInfo.Columns)]
-			datums = datums[:len(idxInfo.Columns)]
+		var handle int64
+		if len(b) > 0 {
+			var handleDatum types.Datum
+			_, handleDatum, err = codec.DecodeOne(b)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			handle = handleDatum.GetInt64()
 		} else {
-			var intHandle int64
-			intHandle, err = decodeHandle(it.Value())
+			handle, err = decodeHandle(it.Value())
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
-			handle.SetInt64(intHandle)
 		}
-		data, err := codec.EncodeValue(nil, datums...)
-		if err != nil {
-			return nil, errors.Trace(err)
+		match, err := rs.evalWhereForRow(ctx, handle, values)
+		if !match {
+			continue
 		}
-		handleData, err := codec.EncodeValue(nil, handle)
-		if err != nil {
-			return nil, errors.Trace(err)
+		if ctx.aggregate {
+			// Update aggregate functions.
+			err = rs.aggregate(ctx, handle, values)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		} else {
+			// If without aggregate functions, just return raw row data.
+			var handleData []byte
+			handleData, err = codec.EncodeValue(nil, types.NewIntDatum(handle))
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			row := &tipb.Row{Handle: handleData}
+			for _, id := range ids {
+				row.Data = append(row.Data, values[id]...)
```

Comment:
No matter a Datum is encoded as key or as value, we can decode it correctly with the flag.

No, we can't change them, index key doesn't need to contain column id before value, but row does.